### PR TITLE
Refactoring and Bug Fixes

### LINKED ISSUE #226: [PE-D][Tester D] Exiting application
Should add that the use of both exit and bye allows exiting from application

-------------
Labels: `severity.VeryLow` `type.DocumentationBug`
original: nyh3/ped#10

### PR DESCRIPTION
**Pull request by nkotaa**
# Description
Changes:
1. Decoupled `UI` from `Parser`
2. General refactor to command parsers
3. Fix crash when negative number of results given to Search
4. Close #226 

To do:
- [x] Update DG to reflect new UI class structure

### CODE REVIEW

File: src/main/java/seedu/binbash/parser/CommandOptionAdder.java [hunk 1]
```diff
@@ -8,16 +8,34 @@
...
+    public ArrayList<ArrayList<OptDesc>> getAllCommandsOptionDescriptions() {
+        return allCommandsOptionDescriptions;
+    }
+
+    /**
+     * Saves option descriptions for the current command after all options are added.
+     *
+     * @param command The name of the command.
+     */
+    void saveCommandOptionDescriptions(String command) {
```
Comment (YHWong20) [inline src/main/java/seedu/binbash/parser/CommandOptionAdder.java:37]: It seems like the `command` parameter is unused.

**Review by YHWong20 — APPROVED** [review]
Looks great! Thanks for removing the direct coupling between `Ui` and `Parser`. Great idea to use `CommandCompleter` as an abstraction layer between the classes as well.

File: src/main/java/seedu/binbash/parser/CommandOptionAdder.java (hunk 1 above)
Comment (nkotaa) [reply src/main/java/seedu/binbash/parser/CommandOptionAdder.java:38]: Now it is.

File: src/main/java/seedu/binbash/parser/SearchCommandParser.java [hunk 2]
```diff
@@ -125,6 +113,9 @@ public SearchCommand parse(String[] commandArgs) throws ParseException {
 
         if (commandLine.hasOption("list")) {
             int numberOfResults = Parser.parseIntOptionValue(commandLine.getOptionValue("list"), "number of results");
+            if (numberOfResults <= 0) {
+                throw new ParseException("number of results must be positive");
+            }
```
Comment (nur-haziq) [inline src/main/java/seedu/binbash/parser/SearchCommandParser.java:118]: Good catch.

**Review by nur-haziq — APPROVED** [review]
LGTM. Thanks for the well formatted PR description 😄

**Pull request merged**